alns/select/MABSelector.py

```diff
@@ -0,0 +1,216 @@
+import itertools
+from typing import Callable, List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from mabwiser.mab import MAB
+from mabwiser.utils import Num
+from numpy.random import RandomState
+
+from alns.State import State
+from alns.select.OperatorSelectionScheme import OperatorSelectionScheme
+
+
+class MABSelector(OperatorSelectionScheme):
+    """
+    A selector that uses any multi-armed-bandit algorithm from MABWiser.
+
+    This selector is a wrapper around the many multi-armed bandit algorithms
+    available in the `MABWiser <https://github.com/fidelity/mabwiser>`_
+    library. Since ALNS operator selection can be framed as a
+    multi-armed-bandit problem, this wrapper allows you to use a variety of
+    existing multi-armed-bandit algorithms as operator selectors instead of
+    having to reimplement them.
+
+    Note that the supplied ``MAB`` object must be generated with the static
+    method ``make_arms``.
+
+    Parameters
+    ----------
+    scores
+        A list of four non-negative elements, representing the rewards when the
+        candidate solution results in a new global best (idx 0), is better than
+        the current solution (idx 1), the solution is accepted (idx 2), or
+        rejected (idx 3).
+    mab
+        A mabwiser MAB object that will be used to select the
+        (destroy, repair) operator pairs. Them arms of the ``mab`` object must
+        be generated with the static method ``make_arms``.
+    num_destroy
+        Number of destroy operators.
+    num_repair
+        Number of repair operators.
+    op_coupling
+        Optional boolean matrix that indicates coupling between destroy and
+        repair operators. Entry (i, j) is True if destroy operator i can be
+        used together with repair operator j, and False otherwise.
+    context_extractor
+        Optional function that takes a State object and returns a context
+        vector for that state that can be passed to a contextual mabwiser
+        bandit. If the MAB algorithm supports it, this context will be used to
+        help predict the next (destroy, repair) combination.
```

Comment:
I get the idea but I am not following how this contextual version is gonna work. Do we have a "running" example to make this concrete so that I can follow how exactly is this gonna work